### ISSUE TITLE
Segmentation fault (core dumped)

### ISSUE THREAD

**Issue by wangyirui** [pytorch/pytorch#6449]
When I training my model, after hundreds of batches, the segmentation fault always occurs. 

- PyTorch or Caffe2: Pytorch 
- OS: Ubuntu 16.04
- PyTorch version: 0.4.0a0+57ee202
- How you installed PyTorch (conda, pip, source): source
- Python version: 3.6
- CUDA/cuDNN version: 9.1

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffe958f5ca in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
(gdb) where
#0  0x00007fffe958f5ca in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#1  0x00007fffe94b5269 in ?? () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#2  0x00007fffe95ecd83 in cuEventDestroy_v2 () from /usr/lib/x86_64-linux-gnu/libcuda.so.1
#3  0x00007fffb3535ec4 in ?? () from /usr/local/cuda/lib64/libcudnn.so.7
#4  0x00007fffb356e3b7 in ?? () from /usr/local/cuda/lib64/libcudnn.so.7
#5  0x00007fffb32989d4 in ?? () from /usr/local/cuda/lib64/libcudnn.so.7
#6  0x00007fffb3295035 in ?? () from /usr/local/cuda/lib64/libcudnn.so.7
#7  0x00007fffb329587e in cudnnRNNForwardInference () from /usr/local/cuda/lib64/libcudnn.so.7
#8  0x00007fffd5a04f47 in at::native::_cudnn_rnn(at::Tensor const&, at::ArrayRef<at::Tensor>, long, at::Tensor const&, at::Tensor const&, at::Tensor const&, long, long, long, bool, double, bool, bool, at::ArrayRef<long>, at::Tensor const&) ()
   from /home/kimi/anaconda2/envs/py36/lib/python3.6/site-packages/torch/lib/libATen.so
#9  0x00007fffd5c706d4 in at::Type::_cudnn_rnn(at::Tensor const&, at::ArrayRef<at::Tensor>, long, at::Tensor const&, at::Tensor const&, at::Tensor const&, long, long, long, bool, double, bool, bool, at::ArrayRef<long>, at::Tensor const&) const ()
   from /home/kimi/anaconda2/envs/py36/lib/python3.6/site-packages/torch/lib/libATen.so
#10 0x00007fffe70636bb in torch::autograd::VariableType::_cudnn_rnn (this=<optimized out>, input=..., 
    weight=..., weight_stride0=<optimized out>, weight_buf=..., hx=..., cx=..., mode=2, hidden_size=1000, 
    num_layers=4, batch_first=false, dropout=<optimized out>, train=false, bidirectional=false, 
    batch_sizes=..., dropout_state=...) at torch/csrc/autograd/generated/VariableType.cpp:17825
#11 0x00007fffe7151599 in at::_cudnn_rnn (dropout_state=..., batch_sizes=..., bidirectional=false, 
    train=false, dropout=0, batch_first=false, num_layers=4, hidden_size=1000, mode=2, cx=..., hx=..., 
    weight_buf=..., weight_stride0=4, weight=..., input=...)
---Type <return> to continue, or q <return> to quit---
    at /home/kimi/Downloads/pytorch/torch/lib/tmp_install/include/ATen/Functions.h:2883
#12 torch::autograd::dispatch__cudnn_rnn (dropout_state=..., batch_sizes=..., bidirectional=false, 
    train=false, dropout=0, batch_first=false, num_layers=4, hidden_size=1000, mode=2, cx=..., hx=..., 
    weight_buf=..., weight_stride0=4, weight=..., input=...)
    at torch/csrc/autograd/generated/python_torch_functions_dispatch.h:309
#13 torch::autograd::THPVariable__cudnn_rnn (self=<optimized out>, args=<optimized out>, 
    kwargs=<optimized out>) at torch/csrc/autograd/generated/python_torch_functions.cpp:578
#14 0x0000555555662fd4 in _PyCFunction_FastCallDict ()
#15 0x00005555556f0bec in call_function ()
#16 0x000055555571519a in _PyEval_EvalFrameDefault ()
#17 0x00005555556ebb6e in PyEval_EvalCodeEx ()
#18 0x00005555556ec456 in function_call ()
#19 0x0000555555662dde in PyObject_Call ()
#20 0x0000555555716994 in _PyEval_EvalFrameDefault ()
#21 0x00005555556e9dfe in _PyEval_EvalCodeWithName ()
#22 0x00005555556eaa11 in fast_function ()
#23 0x00005555556f0cc5 in call_function ()
#24 0x000055555571519a in _PyEval_EvalFrameDefault ()
#25 0x00005555556e99a6 in _PyEval_EvalCodeWithName ()
#26 0x00005555556eb108 in _PyFunction_FastCallDict ()
#27 0x000055555566339f in _PyObject_FastCallDict ()
#28 0x0000555555667ff3 in _PyObject_Call_Prepend ()
#29 0x0000555555662dde in PyObject_Call ()
---Type <return> to continue, or q <return> to quit---
#30 0x0000555555716994 in _PyEval_EvalFrameDefault ()
#31 0x00005555556e99a6 in _PyEval_EvalCodeWithName ()
#32 0x00005555556eaeee in _PyFunction_FastCallDict ()
#33 0x000055555566339f in _PyObject_FastCallDict ()
#34 0x0000555555667ff3 in _PyObject_Call_Prepend ()
#35 0x0000555555662dde in PyObject_Call ()
#36 0x00005555556be901 in slot_tp_call ()
#37 0x00005555556631bb in _PyObject_FastCallDict ()
#38 0x00005555556f0d3e in call_function ()
#39 0x000055555571519a in _PyEval_EvalFrameDefault ()
#40 0x00005555556eae4b in _PyFunction_FastCallDict ()
#41 0x000055555566339f in _PyObject_FastCallDict ()
#42 0x0000555555667ff3 in _PyObject_Call_Prepend ()
#43 0x0000555555662dde in PyObject_Call ()
#44 0x0000555555716994 in _PyEval_EvalFrameDefault ()
#45 0x00005555556e99a6 in _PyEval_EvalCodeWithName ()
#46 0x00005555556eaeee in _PyFunction_FastCallDict ()
#47 0x000055555566339f in _PyObject_FastCallDict ()
#48 0x0000555555667ff3 in _PyObject_Call_Prepend ()
#49 0x0000555555662dde in PyObject_Call ()
#50 0x00005555556be901 in slot_tp_call ()
#51 0x00005555556631bb in _PyObject_FastCallDict ()
#52 0x00005555556f0d3e in call_function ()
---Type <return> to continue, or q <return> to quit---
#53 0x000055555571519a in _PyEval_EvalFrameDefault ()
#54 0x00005555556e9dfe in _PyEval_EvalCodeWithName ()
#55 0x00005555556eaa11 in fast_function ()
#56 0x00005555556f0cc5 in call_function ()
#57 0x000055555571519a in _PyEval_EvalFrameDefault ()
#58 0x00005555556e99a6 in _PyEval_EvalCodeWithName ()
#59 0x00005555556eaa11 in fast_function ()
#60 0x00005555556f0cc5 in call_function ()
#61 0x0000555555715eb1 in _PyEval_EvalFrameDefault ()
#62 0x00005555556e99a6 in _PyEval_EvalCodeWithName ()
#63 0x00005555556eaa11 in fast_function ()
#64 0x00005555556f0cc5 in call_function ()
#65 0x0000555555715eb1 in _PyEval_EvalFrameDefault ()
#66 0x00005555556e9dfe in _PyEval_EvalCodeWithName ()
#67 0x00005555556eaa11 in fast_function ()
#68 0x00005555556f0cc5 in call_function ()
#69 0x000055555571519a in _PyEval_EvalFrameDefault ()
#70 0x00005555556e9dfe in _PyEval_EvalCodeWithName ()
#71 0x00005555556eaa11 in fast_function ()
#72 0x00005555556f0cc5 in call_function ()
#73 0x000055555571519a in _PyEval_EvalFrameDefault ()
#74 0x00005555556eb529 in PyEval_EvalCodeEx ()
#75 0x00005555556ec2cc in PyEval_EvalCode ()
---Type <return> to continue, or q <return> to quit---
#76 0x0000555555768af4 in run_mod ()
#77 0x0000555555768ef1 in PyRun_FileExFlags ()
#78 0x00005555557690f4 in PyRun_SimpleFileExFlags ()
#79 0x000055555576cc28 in Py_Main ()
#80 0x000055555563471e in main ()

**Comment by ailzhang** [comment]
Could you provide a minimal repro script? Thanks!

**Comment by wangyirui** [comment]
@ailzhang 
```
# training process
        for i, sample in enumerate(itr):
            if use_cuda:
                # wrap input tensors in cuda tensors
                sample = utils.make_variable(sample, cuda=cuda)

            with torch.no_grad():

                neg_tokens = translator.generate_translation_tokens(sample, beam_size=args.beam, maxlen_a=args.max_len_a,
                                                                    maxlen_b=args.max_len_b, nbest=args.nbest, max_res=args.fixed_max_len)
                neg_labels = sample['target'].new(neg_tokens.size(0), 1).fill_(0).float()

                pos_tokens = sample['net_input']['src_tokens']

                pos_labels = sample['target'].new(neg_tokens.size(0), 1).fill_(1).float()

                src_tokens = torch.cat([sample['net_input']['src_tokens']]*2, dim=0)
                trg_tokens = torch.cat([pos_tokens, neg_tokens], dim=0)
                labels = torch.cat([pos_labels, neg_labels], dim=0)
                indices = np.random.permutation(labels.size(0))
                src_tokens = src_tokens[indices]
                trg_tokens = trg_tokens[indices]
                labels = labels[indices]

            disc_out = discriminator(src_tokens, trg_tokens, dataset.dst_dict.pad())
            loss = criterion(disc_out, labels)
            # _, prediction = disc_out.topk(1)
            prediction = torch.round(disc_out)
            acc = torch.sum(prediction == labels).float() / len(labels)
            logging_meters['train_acc'].update(acc)
            logging_meters['train_loss'].update(loss)
            logging.debug("D training loss {0:.3f}, acc {1:.3f}, avgAcc {2:.3f}, lr={3} at batch {4}: ".format(logging_meters['train_loss'].avg,
                                                                                                                acc,
                                                                                                               logging_meters['train_acc'].avg,
                                                                                                               optimizer.param_groups[0]['lr'],
                                                                                                               i,))

            optimizer.zero_grad()
            loss.backward()
            torch.nn.utils.clip_grad_norm(discriminator.parameters(), args.clip_norm)
            optimizer.step()
```

**Comment by ailzhang** [comment]
Hi @wangyirui , will you be able to share the whole script as the error show up after a few hundred iterations, it's hard to repro on our side.

**Comment by wangyirui** [comment]
@ailzhang the whole script can be found here: https://github.com/wangyirui/Adversarial-NMT
This error occurs when running function train_d

**Comment by wangyirui** [comment]
@apaszke I am trying to find the minimal part that can reproduce this issue. But it's a large repo so I am not sure which part causes the crash...

**Comment by ezyang** [comment]
CC @ngimel, in case this looks familiar. (Not pointing fingers; I also rewrote this codepath for 0.4 and I could have gotten it wrong.)

**Comment by ngimel** [comment]
Small repro would help, also, what is cudnn version? The error looks like it is coming from inside cudnn, on the destruction of internal event, so it is unlikely that something in pytorch is causing it, but hard to say for sure without a repro.

**Comment by wangyirui** [comment]
Hi all, thanks for your help. After installing the latest master's branch, this problem seems gone...
